fix: Relative imports sometimes generated with backward slashes on Windows

marcelomendoncasoares opened the pull request:
Closes issue #3145.

## Pre-launch Checklist

- [x] I read the [Contribute](https://docs.serverpod.dev/contribute) page and followed the process outlined there for submitting PRs.
- [x] This update contains only one single feature or bug fix and nothing else. (If you are submitting multiple fixes, please make multiple PRs.)
- [x] I read and followed the [Dart Style Guide](https://dart.dev/guides/language/effective-dart/style) and formatted the code with [dart format](https://dart.dev/tools/dart-format).
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`), and made sure that the documentation follows the same style as other Serverpod documentation. I checked spelling and grammar.
- [x] I added new tests to check the change I am making.
- [x] All existing and new tests are passing.
- [x] Any breaking changes are documented below.

If you need help, consider asking for advice on the [discussion board](https://github.com/serverpod/serverpod/discussions).

## Breaking changes

No breaking changes.

Isakdl commented on tools/serverpod_cli/lib/src/generator/dart/library_generators/util/custom_allocators.dart:
```diff
@@ -14,10 +14,11 @@ class ImportCollector {
+
+    // If on Windows, top level paths could appear with backslashes and break
+    // the import clause, such as `import '..\protocol.dart' as _i1;`.
+    return relativePath.replaceAll('\\', '/');
```
Suggestion: The above code will fail if a file is named with '\' with in the file name, change this line to the following instead:

```dart
return p.split(relativePath).join('/');
```

marcelomendoncasoares replied:
You're right, good point. Just changed to your suggestion.

Isakdl reviewed (changes requested):
Thank you for the PR! 

I have made a suggestion to make your solution a little more robust, I would also like to add two tests for this case.

Add a tests in this file: https://github.com/serverpod/serverpod/blob/fc8e4e2a26e7301ae45ec07b98f3b1f7f7993c82/tools/serverpod_cli/test/generator/dart/server_code_generator/sealed_class_test.dart#L18

And copy the test for the client as well, in the sibling folder `client_code_generator`. Currently there is no sealed_class_test.dart file on the client, but create one and only append your test there!

The case to check for is validating that the import to protocol.dart is correct, e.g. the output code should contain `import '../protocol.dart'`

marcelomendoncasoares commented:
> The case to check for is validating that the import to protocol.dart is correct, e.g. the output code should contain `import '../protocol.dart'`

@Isakdl, I added the suggested test, but there is a problem. Although the real code generation have the protocol import - that is used in the part directives for extension of the parent class -, the test code generation is missing it.

The actual parent class on `tests\serverpod_test_server\lib\src\generated\inheritance\sealed_parent.dart` declares the protocol import as `_i1`:
```dart
import '../protocol.dart' as _i1;
import 'package:serverpod/serverpod.dart' as _i2;
part 'sealed_child.dart';
part 'sealed_grandchild.dart';
part 'sealed_other_child.dart';
```

And the actual child class on `tests\serverpod_test_server\lib\src\generated\inheritance\sealed_child.dart` uses it to reference the parent class on the extends:
```dart
part of 'sealed_parent.dart';

class SealedChild extends _i1.SealedParent
    implements _i2.SerializableModel, _i2.ProtocolSerialization {
  SealedChild({
    required super.sealedInt,
    required super.sealedString,
    this.nullableInt,
  });
```

But the `parentCompilationUnit` variable of the tests returns:
> "import 'package:serverpod/serverpod.dart' as _i1; part 'example_child.dart'; sealed class Example implements _i1.SerializableModel, _i1.ProtocolSerialization {Example({required this.name}); String name;}"

And the `childCompilationUnit` extends the class directly:
> "part of 'example.dart'; abstract class ExampleChild extends Example implements _i1.SerializableModel, _i1.ProtocolSerialization {ExampleChild._({required super.name, required this.age}); ...

The child is extending the `Example` class directly, without referencing it from a protocol import.

I can confirm that the relative path fix solves the problem because the `serverpod generate` command no longer breaks the import on the sealed classes. But the added test will fail, so I added it only on the server for now and with a skip clause. Any clues on why the generated code for the test is different from the real one?

Isakdl commented:
Hey @marcelomendoncasoares 

I took a look a bit deeper on this, I don't think the bug is related to sealed classes at all. The problem is with having the model in nested folders. 

The reason that the test is not including the `protocol` import is that the test data is not configured to need this import. This import is needed when referencing another model as a class type.


So first we have to configure the test data to set the file to be inside a folder. This can be done with the method `withSubDirParts` on the `ClassDefinitionBuilder`.

Then we need to configure a class reference to another model as a field on the current class.

After that we will get the `'../protocol.dart'` import.

Full test setup:


```dart
import 'package:serverpod_cli/src/util/model_helper.dart'; // imports defaultModuleAlias

    var example = ClassDefinitionBuilder()
        .withClassName('Example')
        .withFileName('example')
        .withSubDirParts(['folder'])
        .withField(FieldDefinitionBuilder()
            .withName('myClass')
            .withType(
              TypeDefinitionBuilder()
                  .withClassName('RefClass')
                  .withUrl(defaultModuleAlias)
                  .build(),
            )
            .build())
        .build();

    var refClass = ClassDefinitionBuilder()
        .withClassName('RefClass')
        .withFileName('ref_class')
        .withSimpleField('name', 'String')
        .build();
```

But I think this test actually belongs in a new test file, if you could create one that we call `imports_test.dart` and move the test there with this setup! Then copy the entire file into the `client_code_generator` as well. 

E.g. we end up with

- client_code_generator
  - imports_test.dart
- server_code_generator
  - imports_test.dart

Isakdl commented:
@marcelomendoncasoares 

Sorry I made one mistake here it is indeed related to sealed classes.

This is the test construction needed, but I still think it makes sense to put this in a standalone file.

```dart
   var parent = ClassDefinitionBuilder()
        .withClassName('Example')
        .withFileName('example')
        .withField(FieldDefinitionBuilder()
            .withName('user')
            .withType(
              TypeDefinitionBuilder()
                  .withClassName('User')
                  .withUrl(defaultModuleAlias)
                  .build(),
            )
            .build())
        .withIsSealed(true) 
        .build();

    var child = ClassDefinitionBuilder()
        .withClassName('ExampleChild')
        .withFileName('example_child')
        .withSimpleField('age', 'int')
        .withExtendsClass(parent)
        .build();

    var user = ClassDefinitionBuilder()
        .withClassName('User')
        .withFileName('user')
        .withSimpleField('name', 'String')
        .build();
```

marcelomendoncasoares commented:
Thanks for the help, @Isakdl! Just made the suggested changes, with the tests on standalone files.

Isakdl reviewed (approved):
LGTM, thank you for the work on this @marcelomendoncasoares 🚀 🚀 🚀
